ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,297 @@
+from dataclasses import dataclass
+from enum import IntEnum
+import re
+import os
+from typing import Dict, List, Optional, Tuple
+from ofrak.model.resource_model import ResourceIndexedAttribute
+from ofrak.core.filesystem import FilesystemEntry
+from ofrak.model.resource_model import Data
+from ofrak.service.resource_service_i import ResourceAttributeValueFilter, ResourceFilter
+from black import format_str, FileMode
+
+from ofrak.resource import Resource
+
+
+class SelectableAttributesError(Exception):
+    """
+    Prompt the user for an attribute to select with
+    """
+
+
+class ActionType(IntEnum):
+    UNPACK = 0
+    MOD = 1
+    PACK = 2
+    UNDEF = 3
+
+
+@dataclass
+class ScriptAction:
+    """
+    Encapsulates the structure of a single action within the script, which consists of the string
+    representation of the code for that action and the action's type.
+    """
+
+    action_type: ActionType
+    action: str
+
+
+class ScriptSession:
+    """
+    A script, consisting of an ordered sequence of script actions and a mapping between resources
+    and their autogenerated variable names.
+    """
+
+    actions: List[ScriptAction] = []
+    resource_variable_names: Dict[bytes, str] = {}
+    boilerplate_header: str = r"""
+    from ofrak import *
+    from ofrak.core import *
+
+    async def main(ofrak_context: OFRAKContext):
+    """
+    # TODO: Replace with backend in use by OFRAK instance used to create the script.
+    boilerplate_footer: str = r"""
+    if __name__ == "__main__":
+
+        if False:
+            import ofrak_angr
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_angr)
+
+        if False:
+            import ofrak_binary_ninja
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_binary_ninja)
+
+        if False:
+            import ofrak_ghidra
+            ofrak.discover(ofrak_ghidra)
```

Comment:
Running `black` changes this, so we probably want to have the version here match the version `black` outputs.